### PR TITLE
Add API for communicating with Docker volume plugins

### PR DESCRIPTION
**Pull request by mheon**
Docker provides extensibility through a plugin system, of which several types are available. This provides an initial library API for communicating with one type of plugins, volume plugins. Volume plugins allow for an external service to create and manage a volume on Podman's behalf.

This does not integrate the plugin system into Libpod or Podman yet; that will come in subsequent pull requests.

This replaces #4548 and will begin the continuation of that work.

### CODE REVIEW

**Comment by mheon** [comment]
I'm especially interested from a review from @jwhonce on the HTTP bits.

**Comment by mheon** [comment]
@rhatdan will also be interested in this.

File: go.mod [hunk 1]
```diff
@@ -7,13 +7,13 @@ require (
...
 	github.com/buger/goterm v0.0.0-20181115115552-c206103e1f37
 	github.com/checkpoint-restore/go-criu v0.0.0-20190109184317-bdb7599cd87b
-	github.com/codahale/hdrhistogram v0.0.0-20161010025455-3a0bb77429bd // indirect
 	github.com/containernetworking/cni v0.8.0
 	github.com/containernetworking/plugins v0.8.7
 	github.com/containers/buildah v1.17.1-0.20201113135631-d0c958d65eb2
 	github.com/containers/common v0.27.0
 	github.com/containers/conmon v2.0.20+incompatible
 	github.com/containers/image/v5 v5.8.0
+	github.com/containers/libpod v1.9.3
```
Comment (mheon) [inline go.mod:16]: Eeek. I don't know how this got in here...

**Review by jwhonce — COMMENTED** [review]
This is using a seperate HTTP Server from current API server. That is not "bad" but it does imply as written: 
1) people looking at this code need to know this server configuration is not impacted by pkg/api/server, 
2) this is hard-wiring a unix domain socket on the server while the imported library does allow tcp connections

I also didn't catch where the socket is being created/cleaned up

**Comment by mheon** [comment]
This is not a server, it is a client. The servers are run by individual plugins; we connect to the plugins via Unix socket and speak their REST API when we want them to perform services for us. The lifecycle of their sockets is external to Podman and is the business of the plugin creator.

File: libpod/plugin/volume_api.go [hunk 2]
```diff
@@ -0,0 +1,451 @@
...
+
+	req.Header.Set("Host", newPlugin.getURI())
+	req.Header.Set("Content-Type", sdk.DefaultContentTypeV1_1)
+
+	client := new(http.Client)
+	client.Timeout = defaultTimeout
+	resp, err := client.Do(req)
+	if err != nil {
+		return errors.Wrapf(err, "error sending request to plugin %s activation endpoint", newPlugin.Name)
+	}
```
Comment (jwhonce) [inline libpod/plugin/volume_api.go:91]:
```suggestion
	}
	defer resp.Body.Close()
```

File: libpod/plugin/volume_api.go [hunk 3]
```diff
@@ -0,0 +1,451 @@
...
+
+	if err := p.handleErrorResponse(resp, listPath, ""); err != nil {
+		return nil, err
+	}
+
+	// TODO: Can probably unify response reading under a helper
+	volumeRespBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading response body from volume plugin %s", p.Name)
+	}
```
Comment (jwhonce) [inline libpod/plugin/volume_api.go:293]:
```suggestion
	}
	defer resp.Body.Close()
```

**Review by jwhonce — CHANGES_REQUESTED** [review]

File: libpod/plugin/volume_api.go [hunk 4]
```diff
@@ -0,0 +1,453 @@
...
+
+	// Read and decode the body so we can tell if this is a volume plugin.
+	respBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Wrapf(err, "error reading activation response body from plugin %s", newPlugin.Name)
+	}
+
+	respStruct := new(activateResponse)
+	if err := json.Unmarshal(respBytes, respStruct); err != nil {
+		return errors.Wrapf(err, "error unmarshaling plugin %s activation response", newPlugin.Name)
```
Comment (TomSweeneyRedHat) [inline libpod/plugin/volume_api.go:109]:
```suggestion
		return errors.Wrapf(err, "error unmarshalling plugin %s activation response", newPlugin.Name)
```

File: libpod/plugin/volume_api.go [hunk 5]
```diff
@@ -0,0 +1,453 @@
...
+
+	// TODO: Can probably unify response reading under a helper
+	volumeRespBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading response body from volume plugin %s", p.Name)
+	}
+
+	volumeResp := new(volume.ListResponse)
+	if err := json.Unmarshal(volumeRespBytes, volumeResp); err != nil {
+		return nil, errors.Wrapf(err, "error unmarshaling volume plugin %s list response", p.Name)
```
Comment (TomSweeneyRedHat) [inline libpod/plugin/volume_api.go:299]:
```suggestion
		return nil, errors.Wrapf(err, "error unmarshalling volume plugin %s list response", p.Name)
```

File: libpod/plugin/volume_api.go [hunk 6]
```diff
@@ -0,0 +1,453 @@
...
+	}
+
+	getRespBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error reading response body from volume plugin %s", p.Name)
+	}
+
+	getResp := new(volume.GetResponse)
+	if err := json.Unmarshal(getRespBytes, getResp); err != nil {
+		return nil, errors.Wrapf(err, "error unmarshaling volume plugin %s get response", p.Name)
```
Comment (TomSweeneyRedHat) [inline libpod/plugin/volume_api.go:334]:
```suggestion
		return nil, errors.Wrapf(err, "error unmarshalling volume plugin %s get response", p.Name)
```

File: libpod/plugin/volume_api.go [hunk 7]
```diff
@@ -0,0 +1,453 @@
...
+	}
+
+	pathRespBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", errors.Wrapf(err, "error reading response body from volume plugin %s", p.Name)
+	}
+
+	pathResp := new(volume.PathResponse)
+	if err := json.Unmarshal(pathRespBytes, pathResp); err != nil {
+		return "", errors.Wrapf(err, "error unmarshaling volume plugin %s path response", p.Name)
```
Comment (TomSweeneyRedHat) [inline libpod/plugin/volume_api.go:390]:
```suggestion
		return "", errors.Wrapf(err, "error unmarshalling volume plugin %s path response", p.Name)
```

File: libpod/plugin/volume_api.go [hunk 8]
```diff
@@ -0,0 +1,453 @@
...
+	}
+
+	mountRespBytes, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", errors.Wrapf(err, "error reading response body from volume plugin %s", p.Name)
+	}
+
+	mountResp := new(volume.MountResponse)
+	if err := json.Unmarshal(mountRespBytes, mountResp); err != nil {
+		return "", errors.Wrapf(err, "error unmarshaling volume plugin %s path response", p.Name)
```
Comment (TomSweeneyRedHat) [inline libpod/plugin/volume_api.go:427]:
```suggestion
		return "", errors.Wrapf(err, "error unmarshalling volume plugin %s path response", p.Name)
```

**Comment by umohnani8** [comment]
code LGTM once the suggestions are addressed.

**Comment by mheon** [comment]
All comments addressed

**Comment by TomSweeneyRedHat** [comment]
@mheon looks like a gofmt issue with `/libpod/plugin/volume_api.go`

**Comment by mheon** [comment]
Force-pushed with fix

**Comment by mheon** [comment]
@containers/podman-maintainers This is ready

**Comment by rhatdan** [comment]
@mheon I don't think this should go in until we do 3.0.

**Comment by mheon** [comment]
Sure, probably a good call.

**Comment by mheon** [comment]
3.0 is landed, this is ready now @containers/podman-maintainers

**Comment by umohnani8** [comment]
LGTM @baude @rhatdan PTAL

**Pull request merged**